Fixed `PCL_WARNINGS_ARE_ERRORS` option and some compile warnings

ueqri opened the pull request:
Fixed #4547 . 

At present, `PCL_WARNINGS_ARE_ERRORS` won't work if the custom `CMAKE_CXX_FLAGS` value is provided for cmake at the same time. 

So I move the former option out of the IF scope(which checks whether `CMAKE_CXX_FLAGS` is enabled for custom value).

I've tested the commit in container of `pointcloudlibrary/env:20.04` with cmake command from `.ci/azure-pipelines/build/ubuntu.yaml`.

During my local test, I found the key error(occurs many times in compiling) below, so this PR may be **not safe** for CI build currently due to the ON of `PCL_WARNINGS_ARE_ERRORS` in:
https://github.com/PointCloudLibrary/pcl/blob/e1664a4ba63107b2c5ca10bc4137c5a8f73ae0b5/.ci/azure-pipelines/azure-pipelines.yaml#L35
```text
pcl/common/include/pcl/common/impl/io.hpp:139:12: error: 'void* memcpy(void*, const void*, size_t)' copying an object 
 of non-trivial type 'struct pcl::PointNormal' from an array of 'const struct pcl::PointXYZ' [-Werror=class-memaccess]
  139 |     memcpy (&cloud_out[0], &cloud_in[0], cloud_in.size () * sizeof (PointInT));
      |     ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
https://github.com/PointCloudLibrary/pcl/blob/e1664a4ba63107b2c5ca10bc4137c5a8f73ae0b5/common/include/pcl/common/impl/io.hpp#L139

ueqri commented:
Should I fix the error code segments the CI shows, or just turn `-DPCL_WARNINGS_ARE_ERRORS` to OFF temporarily?

kunaltyagi commented:
Those warnings would need to go away first. The main issue is that 2 separate classes are memcpy'd and it's a bit hard to remove that memcpy without C++17's `if constexpr`.

If you can extract this memcpy into a separate templated function and implement class specialization for the classes that cause the warning, that'd solve the issue

ueqri commented:
Many thanks for your explanations :smiley: ! I'm gonna try to handle those warnings and make new commit.

ueqri commented:
> If you can extract this memcpy into a separate templated function and implement class specialization for the classes that cause the warning, that'd solve the issue

I've made it in my last commit, and the previous errors disappear. However, new errors arise:
```text
/pcl/surface/include/pcl/surface/impl/convex_hull.hpp:218:10: error: 'void* memset(void*, int, size_t)' writing to an object of non-trivial type '__gnu_cxx::__alloc_traits<Eigen::aligned_allocator<pcl::PointXYZ>, pcl::PointXYZ>::value_type' {aka 'struct pcl::PointXYZ'}; use assignment instead [-Werror=class-memaccess]
  218 |   memset (&hull.points[0], hull.size (), sizeof (PointInT));
      |   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

/pcl/surface/include/pcl/surface/impl/convex_hull.hpp:225:10: error: 'void* memset(void*, int, size_t)' writing to an object of type '__gnu_cxx::__alloc_traits<Eigen::aligned_allocator<std::pair<int, Eigen::Matrix<float, 4, 1> > >, std::pair<int, Eigen::Matrix<float, 4, 1> > >::value_type' {aka 'struct std::pair<int, Eigen::Matrix<float, 4, 1> >'} with no trivial copy-assignment [-Werror=class-memaccess]
  225 |   memset (&idx_points[0], hull.size (), sizeof (std::pair<int, Eigen::Vector4f>));
      |   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

/pcl/features/include/pcl/features/impl/integral_image2D.hpp:161:10: error: 'void* memset(void*, int, size_t)' clearing an object of type 'using ElementType = class Eigen::Matrix<double, 3, 1>' {aka 'class Eigen::Matrix<double, 3, 1>'} with no trivial copy-assignment; use assignment or value-initialization instead [-Werror=class-memaccess]
  161 |   memset (previous_row, 0, sizeof (ElementType) * (width_ + 1));
      |   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

/pcl/features/include/pcl/features/impl/integral_image2D.hpp:192:12: error: 'void* memset(void*, int, size_t)' clearing an object of type 'using SecondOrderType = class Eigen::Matrix<double, 6, 1, 0, 6, 1>' {aka 'class Eigen::Matrix<double, 6, 1, 0, 6, 1>'} with no trivial copy-assignment; use assignment or value-initialization instead [-Werror=class-memaccess]
  192 |     memset (so_previous_row, 0, sizeof (SecondOrderType) * (width_ + 1));
      |     ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
They are quit the same problems. Here are my solutions. Could you please have a review on it? Full compile log is also pasted [here](https://paste.ubuntu.com/p/jyDHsntvMF/).

### Solutions
For the first two errors, I've found L218 & L225 seem to make no sense in the context, because `hull[0], idx_points[0]` are overridden in this loop(since `i` begins with 0), so I deleted them.
https://github.com/PointCloudLibrary/pcl/blob/fcc948a10d1fe653fb93f4274c155fa38a4b14fe/surface/include/pcl/surface/impl/convex_hull.hpp#L217-L232

For the next two errors, L161 seems unnecessary either, due to the `setZero()` methods in L173. and I think replacing L161 with `previous_row [0].setZero ();` will work equally and handle the error. L192 is same too.
https://github.com/PointCloudLibrary/pcl/blob/fcc948a10d1fe653fb93f4274c155fa38a4b14fe/features/include/pcl/features/impl/integral_image2D.hpp#L155-L221

### BTW
I also found `first_order_integral_image_` (used by `previous_row`, `current_row`, etc.) is `std::vector` type. Iterating using pointer currently e.g. `ElementType* previous_row`, can be changed into some standard methods e.g. `std::iterator`. It's something for another PR.
https://github.com/PointCloudLibrary/pcl/blob/3fab805ae749dc5e1698720dbfad6d3dec4d6c15/features/include/pcl/features/integral_image2D.h#L216-L217

I compile again and something new appears :joy: , not bad, I'll tackle them all. It's exciting for me.:satisfied:

kunaltyagi commented on common/include/pcl/common/impl/io.hpp:
```diff
@@ -136,7 +151,7 @@ copyPointCloud (const pcl::PointCloud<PointInT> &cloud_in,
   if (isSamePointType<PointInT, PointOutT> ())
     // Copy the whole memory block
-    memcpy (&cloud_out[0], &cloud_in[0], cloud_in.size () * sizeof (PointInT));
+    copyPointCloudMemcpy (cloud_in, cloud_out);
```
This one doesn't need shenanigans. You can use `std::copy` here

ueqri replied:
I'm sorry that if I use `std::copy(&cloud_in[0], &(cloud_in[0])+cloud_in.size(), &cloud_out[0]);` to replace the old `memset`, the error is that, this sentence will be definitely compiled regardless of the completely difference of  `PointInT` and `PointOutT`, which may cause error when copy those elements.

Perhaps I didn't catch what you means well about using `std::copy`. Sorry for that. :cry:

In my opinion, if I extract the `memcpy` into a templated function, when `PointInT` and `PointOutT` are different types(which also means `isSamePointType<PointInT, PointOutT> ()` returns 0 during compile-time), the call of `copyPointCloudMemcpy` after its template instantiation will do nothing, which will handle the previous warnings.

kunaltyagi replied:
> this sentence will be definitely compiled regardless of the completely difference of PointInT and PointOutT

Oh ofc. This isn't `if constexpr`.

In that case, what if we replace the memcpy in the `copyPointCloudMemcpy` function?

ueqri replied:
Yeah, I think it will be better if we replace the C-style memcpy with modern C++ method.:satisfied:

mvieth replied:
Not sure if `std::copy` is really the best option here. We are hoping that `std::copy` uses memcpy internally (for speed reasons), so why not keep using memcpy directly? Also: how about moving the for-loop below into the currently empty `copyPointCloudMemcpy` and remove this if-else completely?

ueqri replied:
Thanks for your reviews! :blush:

> why not keep using memcpy directly?

I think `std::copy` is more flexible than `memcpy` for no performance loss, and is safer and increases in speed especially for those non-POD types like `vector<...> points;` in `pcl::PointCloud`. I've investigated it and found this page really helpful: https://stackoverflow.com/questions/4707012/is-it-better-to-use-stdmemcpy-or-stdcopy-in-terms-to-performance

> how about moving the for-loop below into the currently empty copyPointCloudMemcpy and remove this if-else completely?

Excellent idea! :grinning: And make the code more readable and prettier. Thank you! I'll fix it then.

kunaltyagi replied:
> We are hoping that std::copy uses memcpy internally (for speed reasons), so why not keep using memcpy directly?
> observations from SO thread and `std::copy` being more flexible/faster than `memcpy`

* `std::copy` can handle types which might have non-trivial copy ctors (PCL supports custom points. Using `copy` allows us to support wider cadre of structs as points, not a big priority to-be-honest)
* depending on the distribution and C stdlib, `memcpy` might not be inlined (since it's a pure C function), while `std::copy` has a higher chance of being inlined (if they don't use `memcpy` themselves, due to being templated)
* Some hardened distributions don't use sse, avx in their C stdlib, so even if `std::copy` is just a `for(auto& a: b){*a = *c; a++;c++;}`, the application might be faster with `copy` than with `memcpy` (if the optimizations are turned on) (this is rare, so essentially, I just have 1 point 😆 )

[Link to copy in libcxx](https://github.com/llvm/llvm-project/blob/main/libcxx/include/algorithm#L1701) (clang's stdlib for c++)
[GCC's implementation](https://github.com/gcc-mirror/gcc/blob/master/libstdc%2B%2B-v3/include/bits/stl_algobase.h#L338) is similar but way more complicated (relying on combination of memmove, memcpy, simple loop, etc.)

kunaltyagi commented on surface/include/pcl/surface/impl/convex_hull.hpp:
```diff
@@ -215,14 +215,12 @@ pcl::ConvexHull<PointInT>::performReconstruction2D (PointCloud &hull, std::vecto
 
   int num_vertices = qh num_vertices;
   hull.resize (num_vertices);
-  memset (&hull.points[0], hull.size (), sizeof (PointInT));
```
Perhaps, perform:
```cpp
hull.clear()
hull.resize(num_vertices, PointInT{});  // should make them of correct default values (0 or non-0)
```

kunaltyagi commented on common/include/pcl/common/impl/io.hpp:
```diff
@@ -118,6 +118,21 @@ getFieldsList (const pcl::PointCloud<PointT> &)
 }
 
 
+template <typename PointInT, typename PointOutT> static void
+copyPointCloudMemcpy (const pcl::PointCloud<PointInT> &,
+                pcl::PointCloud<PointOutT> &)
+{
+}
+
+
+template <typename PointT> static void
+copyPointCloudMemcpy (const pcl::PointCloud<PointT> &cloud_in,
+                pcl::PointCloud<PointT> &cloud_out)
+{
+  memcpy (&cloud_out[0], &cloud_in[0], cloud_in.size () * sizeof (PointT));
+}
+
+
```
These changes turned out to be unneeded.

kunaltyagi commented:
```
/pcl/surface/include/pcl/surface/impl/convex_hull.hpp:218:10: error: 'void* memset(void*, int, size_t)' writing to an object of non-trivial type '__gnu_cxx::__alloc_traits<Eigen::aligned_allocator<pcl::PointXYZ>, pcl::PointXYZ>::value_type' {aka 'struct pcl::PointXYZ'}; use assignment instead [-Werror=class-memaccess]
  218 |   memset (&hull.points[0], hull.size (), sizeof (PointInT));
      |   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
We can use `tmp=hull.size(); hull.clear();hull.resize(tmp);` here


```
/pcl/features/include/pcl/features/impl/integral_image2D.hpp:161:10: error: 'void* memset(void*, int, size_t)' clearing an object of type 'using ElementType = class Eigen::Matrix<double, 3, 1>' {aka 'class Eigen::Matrix<double, 3, 1>'} with no trivial copy-assignment; use assignment or value-initialization instead [-Werror=class-memaccess]
  161 |   memset (previous_row, 0, sizeof (ElementType) * (width_ + 1));
      |   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
`ElementType::Zero(width_+1)` might work here

```
/pcl/features/include/pcl/features/impl/integral_image2D.hpp:192:12: error: 'void* memset(void*, int, size_t)' clearing an object of type 'using SecondOrderType = class Eigen::Matrix<double, 6, 1, 0, 6, 1>' {aka 'class Eigen::Matrix<double, 6, 1, 0, 6, 1>'} with no trivial copy-assignment; use assignment or value-initialization instead [-Werror=class-memaccess]
  192 |     memset (so_previous_row, 0, sizeof (SecondOrderType) * (width_ + 1));
```
Similar here

ueqri commented:
I'll fix that soon, thanks for your advice.

ueqri commented on surface/src/3rdparty/opennurbs/opennurbs_3dm_settings.cpp:
```diff
@@ -756,6 +754,8 @@ void ON_3dmAnnotationSettings::Default()
   m_bEnableHatchScaling = 1;
+
+  memset(m_reserved, 0, sizeof(m_reserved));
 }
```
I compare all variables in this `Default` function, and make sure all member variables are set to default values. Except existed assignments, I add `memset` for `m_reserved` referred to these in other `Default` functions:
https://github.com/PointCloudLibrary/pcl/blob/6555d8a6f10f0d2fa6f60e99e10360141ad57a89/surface/src/3rdparty/opennurbs/opennurbs_3dm_settings.cpp#L444
https://github.com/PointCloudLibrary/pcl/blob/6555d8a6f10f0d2fa6f60e99e10360141ad57a89/surface/src/3rdparty/opennurbs/opennurbs_3dm_settings.cpp#L485

ueqri commented on surface/src/3rdparty/opennurbs/opennurbs_archive.cpp:
```diff
@@ -4194,7 +4194,14 @@ CUserDataHeaderInfo::CUserDataHeaderInfo()
 
 void CUserDataHeaderInfo::Initialize()
 {
-  memset(this,0,sizeof(*this));
+  m_classid = {};
+  m_itemid = {};
+  m_appid = {};
+  m_3dm_version = 0;
+  m_3dm_opennurbs_version = 0;
+  m_copycount = 0;
+  m_bLastSavedAsGoo = false;
+  m_xform.Zero();
 }
```
I've initialized all member variables of `CUserDataHeaderInfo` to 0. And `m_classid, m_itemid, m_appid` are all `class ON_UUID` without customized constructors, so I set them using zero initialization.

https://github.com/PointCloudLibrary/pcl/blob/6555d8a6f10f0d2fa6f60e99e10360141ad57a89/surface/src/3rdparty/opennurbs/opennurbs_archive.cpp#L4173-L4188

ueqri commented:
Fixed all the previous errors, as well as some similar `memset` errors in new files. (noticing those code are 3rd party, I'm not sure whether it's ok).

Compiling brings new errors now, and I think this kind of `deprecated-declarations` errors are not easy to be solved right now.

Should we just add `-Wno-error=deprecated-declarations` together with `-Werror` for compile?
```text
pcl/surface/src/3rdparty/opennurbs/opennurbs_archive.cpp:15329:56: error: 'int readdir_r(DIR*, dirent*, dirent**)' is deprecated [-Werror=deprecated-declarations]
15329 |     int readdir_errno = readdir_r(m_dir, &m_dirent, &dp);
      |       
```

kunaltyagi commented:
In the 3rd party sources, I'd prefer to disable warnings using pragma instead of modifying the source code. Widespread `-Wno-error=deprecated-declarations` will also allow actual errors to sneak by

ueqri commented:
Thanks for your help! I'd restore last changes and add `pragma` into those files respectively to disable warnings:smile:.

kunaltyagi commented:
For OpenNURBS, you can also add the `-Wno-deprecated...` flag to the CMakeLists.txt in the surface module

ueqri commented:
It's absolutely a better way not modifying 3rd party source codes.:satisfied: Thank you!

ueqri commented on features/include/pcl/features/impl/integral_image2D.hpp:
```diff
@@ -189,7 +189,7 @@ IntegralImage2D<DataType, Dimension>::computeIntegralImages (
     SecondOrderType* so_previous_row = &second_order_integral_image_[0];
     SecondOrderType* so_current_row  = so_previous_row + (width_ + 1);
-    memset (so_previous_row, 0, sizeof (SecondOrderType) * (width_ + 1));
+    *so_previous_row = ElementType::Zero(width_ + 1);
```
I'm sorry for my unfamiliarity with Eigen library. I made these errors with `pcl/features/include/pcl/features/impl/integral_image2D.hpp:192:22` when building. 

What else I should do to solve the errors? Thank you!
```text
/usr/include/eigen3/Eigen/src/Core/AssignEvaluator.h: In instantiation of 'void Eigen::internal::call_assignment_no_alias(Dst&, const Src&, const Func&) [with Dst = Eigen::Matrix<double, 6, 1, 0, 6, 1>; Src = Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, Eigen::Matrix<double, 3, 1> >; Func = Eigen::internal::assign_op<double, double>]':
/usr/include/eigen3/Eigen/src/Core/AssignEvaluator.h:804:27:   required from 'void Eigen::internal::call_assignment(Dst&, const Src&, const Func&, typename Eigen::internal::enable_if<(! Eigen::internal::evaluator_assume_aliasing<Src>::value), void*>::type) [with Dst = Eigen::Matrix<double, 6, 1, 0, 6, 1>; Src = Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, Eigen::Matrix<double, 3, 1> >; Func = Eigen::internal::assign_op<double, double>; typename Eigen::internal::enable_if<(! Eigen::internal::evaluator_assume_aliasing<Src>::value), void*>::type = void*]'
/usr/include/eigen3/Eigen/src/Core/AssignEvaluator.h:782:18:   required from 'void Eigen::internal::call_assignment(Dst&, const Src&) [with Dst = Eigen::Matrix<double, 6, 1, 0, 6, 1>; Src = Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, Eigen::Matrix<double, 3, 1> >]'
/usr/include/eigen3/Eigen/src/Core/PlainObjectBase.h:714:32:   required from 'Derived& Eigen::PlainObjectBase<Derived>::_set(const Eigen::DenseBase<OtherDerived>&) [with OtherDerived = Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, Eigen::Matrix<double, 3, 1> >; Derived = Eigen::Matrix<double, 6, 1, 0, 6, 1>]'
/usr/include/eigen3/Eigen/src/Core/Matrix.h:225:24:   required from 'Eigen::Matrix<_Scalar, _Rows, _Cols, _Options, _MaxRows, _MaxCols>& Eigen::Matrix<_Scalar, _Rows, _Cols, _Options, _MaxRows, _MaxCols>::operator=(const Eigen::DenseBase<OtherDerived>&) [with OtherDerived = Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, Eigen::Matrix<double, 3, 1> >; _Scalar = double; int _Rows = 6; int _Cols = 1; int _Options = 0; int _MaxRows = 6; int _MaxCols = 1]'
pcl/features/include/pcl/features/impl/integral_image2D.hpp:192:22:   required from 'void pcl::IntegralImage2D<DataType, Dimension>::computeIntegralImages(const DataType*, unsigned int, unsigned int) [with DataType = float; unsigned int Dimension = 3]'
pcl/features/include/pcl/features/impl/integral_image2D.hpp:67:3:   required from 'void pcl::IntegralImage2D<DataType, Dimension>::setInput(const DataType*, unsigned int, unsigned int, unsigned int, unsigned int) [with DataType = float; unsigned int Dimension = 3]'
pcl/features/include/pcl/features/impl/integral_image_normal.hpp:133:23:   required from 'void pcl::IntegralImageNormalEstimation<PointInT, PointOutT>::initCovarianceMatrixMethod() [with PointInT = pcl::PointXYZ; PointOutT = pcl::Normal]'
pcl/features/src/integral_image_normal.cpp:46:3:   required from here
/usr/include/eigen3/Eigen/src/Core/AssignEvaluator.h:833:3: error: static assertion failed: YOU_MIXED_MATRICES_OF_DIFFERENT_SIZES
  833 |   EIGEN_STATIC_ASSERT_SAME_MATRIX_SIZE(ActualDstTypeCleaned,Src)
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

kunaltyagi replied:
so_previous_row is of type `SecondOrderType`. And you're assigning it to `ElementType`. It's just a matrix size mismatch

ueqri replied:
I fixed it. Thanks! :satisfied: I'll pay more attention to the code context next time.

ueqri commented:
> For OpenNURBS, you can also add the -Wno-deprecated... flag to the CMakeLists.txt in the surface module

I added some exclusive rules for -Werror only for 3rd party _OpenNURBS_ compile, in the CMakeList of module `surface`.

And the error logs associated with each `Wno-error=*` are listed blow:

#### For `-Wno-error=deprecated-declarations`
```text
pcl/surface/src/3rdparty/opennurbs/opennurbs_archive.cpp:15329:56: error: 'int readdir_r(DIR*, dirent*, dirent**)' is deprecated [-Werror=deprecated-declarations]
15329 |     int readdir_errno = readdir_r(m_dir, &m_dirent, &dp);
      |       
In file included from pcl/surface/include/pcl/surface/3rdparty/opennurbs/opennurbs_system.h:259,
                 from pcl/surface/include/pcl/surface/3rdparty/opennurbs/opennurbs.h:31,
                 from pcl/surface/src/3rdparty/opennurbs/opennurbs_archive.cpp:17:
/usr/include/dirent.h:183:12: note: declared here
  183 | extern int readdir_r (DIR *__restrict __dirp,
      |            ^~~~~~~~~
```

#### For `-Wno-error=class-memaccess`
```text
pcl/surface/src/3rdparty/opennurbs/opennurbs_bounding_box.cpp: In constructor 'ON_ClippingRegion::ON_ClippingRegion()':
pcl/surface/src/3rdparty/opennurbs/opennurbs_bounding_box.cpp:150:30: error: 'void* memset(void*, int, size_t)' clearing an object of non-trivial type 'class ON_ClippingRegion'; use assignment or value-in itialization instead [-Werror=class-memaccess]
  150 |   memset(this,0,sizeof(*this));
      |                              ^
In file included from pcl/surface/include/pcl/surface/3rdparty/opennurbs/opennurbs.h:58,
                 from pcl/surface/src/3rdparty/opennurbs/opennurbs_bounding_box.cpp:17:
pcl/surface/include/pcl/surface/3rdparty/opennurbs/opennurbs_xform.h:766:16: note: 'class ON_ClippingRegion' declared here
  766 | class ON_CLASS ON_ClippingRegion
      |                ^~~~~~~~~~~~~~~~~
```

#### For `-Wno-error=deprecated-copy`
```text
pcl/surface/src/3rdparty/opennurbs/opennurbs_object_history.cpp: In copy constructor 'ON_BoolValue::ON_BoolValue(const ON_BoolValue&)':                                                                       
pcl/surface/src/3rdparty/opennurbs/opennurbs_object_history.cpp:144:7: error: implicitly-declared 'constexpr ON_Value::ON_Value(const ON_Value&)' is deprecated [-Werror=deprecated-copy]                     
  144 | class ON_BoolValue : public ON_Value                                                                                                                                                                             
      |       ^~~~~~~~~~~~                                                                                                                                                                                               
pcl/surface/src/3rdparty/opennurbs/opennurbs_object_history.cpp:88:13: note: because 'ON_Value' has user-provided 'ON_Value& ON_Value::operator=(const ON_Value&)'
   88 |   ON_Value& operator=(const ON_Value&);
      |             ^~~~~~~~
pcl/surface/src/3rdparty/opennurbs/opennurbs_object_history.cpp: In member function 'virtual ON_Value* ON_BoolValue::Duplicate() const':
pcl/surface/src/3rdparty/opennurbs/opennurbs_object_history.cpp:183:32: note: synthesized method 'ON_BoolValue::ON_BoolValue(const ON_BoolValue&)' first required here
  183 |   return new ON_BoolValue(*this);
      |                                ^
```

#### For `-Wno-error=uninitialized`
```text
pcl/surface/src/3rdparty/opennurbs/opennurbs_point.cpp: In constructor 'ON_2dPoint::ON_2dPoint(const ON_4fPoint&)':                                                                                           
pcl/surface/src/3rdparty/opennurbs/opennurbs_point.cpp:3683:5: error: '*<unknown>.ON_2dPoint::x' is used uninitialized in this function [-Werror=uninitialized]                                               
 3683 |   x *= w*h.x;                                                                                                                                                                                                    
      |   ~~^~~~~~~~                                                                                                                                                                                                     
pcl/surface/src/3rdparty/opennurbs/opennurbs_point.cpp:3684:5: error: '*<unknown>.ON_2dPoint::y' is used uninitialized in this function [-Werror=uninitialized]                                               
 3684 |   y *= w*h.y;                                                                                                                                                                                                    
      |   ~~^~~~~~~~ 
```

#### For `-Wno-error=parentheses`
```text
pcl/surface/src/3rdparty/opennurbs/opennurbs_string.cpp: In member function 'ON_String ON_String::Mid(int, int) const':
pcl/surface/src/3rdparty/opennurbs/opennurbs_string.cpp:1122:12: error: unnecessary parentheses in declaration of 's' [-Werror=parentheses]
 1122 |   ON_String(s);
      |            ^
```

ueqri commented:
Also fixed some [deprecated copy](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=58407) for C++11(and later):
```text
pcl/apps/in_hand_scanner/src/integration.cpp:263:40: warning: implicitly-declared 'constexpr pcl::ihs::PointIHS& pcl::ihs::PointIHS::operator=(const pcl::ihs::PointIHS&)' is deprecated [-Werror=deprecated-copy]
  263 |         pt_d_t = PointIHS (pt_d, weight);
      |                                        ^ 
In file included from pcl/apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/common_types.h:60,
                 from pcl/apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/integration.h:45,
                 from pcl/apps/in_hand_scanner/src/integration.cpp:41:
pcl/apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/impl/common_types.hpp:83:14: note: because 'pcl::ihs::PointIHS' has user-provided 'pcl::ihs::PointIHS::PointIHS(const pcl::ihs::PointIHS&)'
   83 |       inline PointIHS (const PointIHS& other)
      |              ^~~~~~~~/
```

```text
pcl/apps/point_cloud_editor/src/transformCommand.cpp: In constructor 'TransformCommand::TransformCommand(const ConstSelectionPtr&, CloudPtr, const float*, float, float, float)':
pcl/apps/point_cloud_editor/src/transformCommand.cpp:55:70: error: implicitly-declared 'Selection::Selection(const Selection&)' is deprecated [-Werror=deprecated-copy]
   55 |   internal_selection_ptr_ = SelectionPtr(new Selection(*selection_ptr));
      |                                                                      ^
In file included from pcl/apps/point_cloud_editor/src/transformCommand.cpp:42:
pcl/apps/point_cloud_editor/include/pcl/apps/point_cloud_editor/selection.h:67:5: note: because 'Selection' has user-provided 'Selection& Selection::operator=(const Selection&)'
   67 |     operator= (const Selection& selection);
      |     ^~~~~~~~
```

ueqri commented:
I borrow CMake config from `.ci/azure-pipelines/azure-pipelines.yaml` & `.ci/azure-pipelines/build/ubuntu.yaml`, based on pointcloudlibrary/env:20.10.

The compile result looks fine without any errors, when open `PCL_WARNINGS_ARE_ERRORS` flag:
https://paste.ubuntu.com/p/9zwVJnJXNS/

Could you take a look again please? Thank you! :satisfied:

kunaltyagi commented on apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/impl/common_types.hpp:
```diff
@@ -80,23 +80,31 @@ namespace pcl
       }
 
-      inline PointIHS (const PointIHS& other)
+      inline PointIHS& operator=(const PointIHS& other)
```
Could we just `default` it or skip mentioning it at all?

ueqri replied:
I think setting the copy assignment function to `default` is a quite better choice, although refactoring those functions works equally. I fixed it in https://github.com/PointCloudLibrary/pcl/pull/4695/commits/00e2f43d8b8494ea2cf954761fb8c95c94fb7b7c. :smiley:

kunaltyagi commented on apps/point_cloud_editor/include/pcl/apps/point_cloud_editor/selection.h:
```diff
@@ -60,6 +60,15 @@ class Selection : public Statistics
+    /// @brief Copy constructor.
+    /// @param selection a const reference to a selection object whose
+    /// properties will be copied.
+    Selection (const Selection& selection)
```
Would we be able to default this as well?

ueqri replied:
Yeah, I fixed in https://github.com/PointCloudLibrary/pcl/pull/4695/commits/00e2f43d8b8494ea2cf954761fb8c95c94fb7b7c. This change is pretty concise than the old one!

kunaltyagi commented on surface/include/pcl/surface/impl/convex_hull.hpp:
```diff
@@ -214,15 +214,16 @@ pcl::ConvexHull<PointInT>::performReconstruction2D (PointCloud &hull, std::vecto
   vertexT * vertex;
   int i = 0;
 
   std::vector<std::pair<int, Eigen::Vector4f>, Eigen::aligned_allocator<std::pair<int, Eigen::Vector4f> > > idx_points (num_vertices);
   idx_points.resize (hull.size ());
```
This is quite confusing. We allocate num_vertices, and resize to hull.size

Is there any relation between these 2 numbers (one is always >, < or == another)?

ueqri replied:
I walked through the relevant sources codes `surface/include/pcl/surface/impl/convex_hull.hpp` `surface/include/pcl/surface/convex_hull.h` `test/surface/test_convex_hull.cpp` and had better understanding about the algorithms and function calls. 

I think, in 2D convex hull reconstruction, the `hull.size()` is exactly the same with `num_vertices` after qhull finds all vertices of convex hull successfully. And it also equals to `idx_points.size()` (which used as a temporary variable for sorting the vertices), and `polygons[0].vertices.size()` (which only use 1 element in 2D reconstruction).

ueqri replied:
And I was wondering if I could changes the code here to:
https://github.com/PointCloudLibrary/pcl/blob/87f1d83377abbbc29b48b268960cdb96e02d327b/surface/include/pcl/surface/impl/convex_hull.hpp#L216-L221
```c++
int num_vertices = qh num_vertices;
hull.clear();
hull.resize(num_vertices);
```
Thus, `hull.size()` would get the same value as `num_vertices`, and be used in the next context.

mvieth replied:
Not sure if the `hull.clear()` is necessary, but else it looks good

ueqri replied:
Perhaps `hull` was used and stored other data inside before passing to this function? I'm not sure about this.😆

kunaltyagi replied:
Thanks for the investigation. In that case,
```suggestion
  AlignedVector<std::pair<int, Eigen::Vector4f>> idx_points (num_vertices);
```
where the following can go in `common/pcl/types.h` (or not. I just find it to be too verbose)
```cpp
template <T>
AlignedVector = std::vector<T, Eigen::aligned_allocator<T>>;
```

ueqri replied:
Using defined data type makes the code pretty clean! Thank you!

kunaltyagi reviewed (commented):
Pretty small change-set for what I thought would be a deep surgery. 🙇🏾

mvieth commented:
Regarding Ubuntu 18.04 GCC: The noexcept-type problem should be fixable by either wrapping `strtof` and `strtod` in a lambda, similar how `strtol` and `strtoll` are wrapped, or call parse_generic like this: `parse_generic<double, double(*)(const char*, char**)>(strtod, argc, argv, str, val)`, similar for the `float` variant. I wasn't able to test this, though. For discussion, [see here](https://stackoverflow.com/questions/46798456/handling-gccs-noexcept-type-warning).

ueqri commented:
Thanks for reminding me! :grinning: I'll test the changes in Ubuntu 18.04 and try to fix the issues.

kunaltyagi commented on common/include/pcl/types.h:
```diff
@@ -49,6 +49,8 @@
 
 #include <cstdint>
 
+#include <Eigen/StdVector>
```
aligned_allocator is in `Eigen/Core`

ueqri replied:
I'm sorry that I just mistaked the include library for `Eigen/StdVector` from [this page](https://eigen.tuxfamily.org/dox-devel/group__TopicStlContainers.html).
> In practice you must use the Eigen::aligned_allocator (not another aligned allocator), and #include <Eigen/StdVector>.

I'll fixed it right now. Thank you very much!

larshg commented on visualization/CMakeLists.txt:
```diff
@@ -61,6 +61,12 @@ if(VTK_RENDERING_BACKEND_OPENGL_VERSION VERSION_LESS 2)
+
+  # Code in this module may use deprecated declarations when using OpenGLv1
+  # Add the GCC exclusive rules for -Werror only for OpenGLv1 compile to avoid build interruption
+  if(CMAKE_COMPILER_IS_GNUCXX)
```
Missing check for OpenGL version here?

ueqri replied:
Thanks for your reviewing!
https://github.com/PointCloudLibrary/pcl/blob/b5d806c0b6b214d23c0d3cb486714175e07e0852/visualization/CMakeLists.txt#L59-L70
I put this block in `if` scope, and L59 will do the check.

ueqri commented:
> Regarding Ubuntu 18.04 GCC: The noexcept-type problem should be fixable by either wrapping `strtof` and `strtod` in a lambda, similar how `strtol` and `strtoll` are wrapped, or call parse_generic like this: `parse_generic<double, double(*)(const char*, char**)>(strtod, argc, argv, str, val)`, similar for the `float` variant. I wasn't able to test this, though. For discussion, [see here](https://stackoverflow.com/questions/46798456/handling-gccs-noexcept-type-warning).

Fixed this issue in https://github.com/PointCloudLibrary/pcl/pull/4695/commits/b5d806c0b6b214d23c0d3cb486714175e07e0852.

And added some exclusive rules for `-Werror` only for OpenGLv1 compile to avoid build interruption.

Now, this branch can be compiled successfully with `-DPCL_WARNINGS_ARE_ERRORS=ON` in Ubuntu 18.04 GCC & Ubuntu 18.04 GCC, and pass all unit tests.

kunaltyagi commented on common/src/parse.cpp:
```diff
@@ -181,14 +181,16 @@ parse_argument (int argc, const char * const * argv, const char * str, T &val) n
 pcl::console::parse_argument (int argc, const char * const * argv, const char * str, double &val)
 {
-  return parse_generic(strtod, argc, argv, str, val);
+  const auto strtod_l = [](const char *str, char **str_end){ return strtod(str, str_end); };
```
Comment for the reason

ueqri replied:
Thanks for your advice! I'll fix it right now! :grinning:

ueqri commented:
I've compiled with `-DPCL_WARNINGS_ARE_ERRORS=ON` on GPU. 

Here are 5 types of warnings which would turn to error and interrupt the compile: 

### Type 1: Used internal header
It's a type of deprecation warning in Eigen(3.3.7)/Core, refered to: https://gitlab.kitware.com/vtk/vtk/-/issues/17661

Could I add `-Wno-error=cpp` for GPU code compile to skip the error? BTW, I wonder if there's a plan to move Eigen to the latest stable release 3.3.9, and I'd like to try the migration :grinning:.
<details>
<summary>detailed logs</summary>

```text
In file included from /usr/local/include/eigen3/Eigen/Core:257:0,
                 from pcl/common/include/pcl/memory.h:48,
                 from pcl/cuda/common/include/pcl/cuda/point_cloud.h:42,
                 from pcl/cuda/io/src/host_device.cu:38:
/usr/local/cuda/include/host_defines.h:54:2: warning: #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead." [-Wcpp]
 #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead."
  ^~~~~~~
```

</details>

### Type 2: Strict aliasing rules

I'd prefer to add `-fno-strict-aliasing` for compile just as the default `CMAKE_CXX_FLAGS` in CMakeLists：
https://github.com/PointCloudLibrary/pcl/blob/64bf08bc88f9a78324f7b7956ece0c111236ac2d/CMakeLists.txt#L107

<details>
<summary>detailed logs</summary>

```text 
pcl/gpu/kinfu/src/internal.h: In function 'bool pcl::device::valid_host(float)':
pcl/gpu/kinfu/src/internal.h:391:43: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
       return *reinterpret_cast<int*>(&value) != 0x7fffffff; //QNAN
                                           ^
```

</details>

### Type 3: Comparison between `signed` and `unsigned` integer

Since signed integer will be casted to unsigned when comparing two values, how about declaring `xIdx` `yIdx` type as `unsigned` to avoid warnings?
https://github.com/PointCloudLibrary/pcl/blob/64bf08bc88f9a78324f7b7956ece0c111236ac2d/cuda/io/src/debayering.cu#L116-L118
<details>
<summary>detailed logs</summary>

```text
pcl/cuda/io/src/debayering.cu:126:28: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
      if (xIdx == 0 || xIdx == width-1 || yIdx == 0 || yIdx == height-1)
                      ~~~~~~^~~~~~~~~~~~~~~~~~~
pcl/cuda/io/src/debayering.cu:126:74: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
      if (xIdx == 0 || xIdx == width-1 || yIdx == 0 || yIdx == height-1)
                                                                    ~~~~  ^     
```

</details>

### Type 4: Deprecated declarations

The warning is about using deprecated function in test codes:
https://github.com/PointCloudLibrary/pcl/blob/64bf08bc88f9a78324f7b7956ece0c111236ac2d/test/gpu/octree/perfomance.cpp#L227
Could I just add exclusive rule like `-Wno-error=deprecated-declarations` in `global_test`(or exactly `test/gpu/`) CMakeLists, or fix the test code with new fuction call?

<details>
<summary>deprecated API & new API</summary>

https://github.com/PointCloudLibrary/pcl/blob/64bf08bc88f9a78324f7b7956ece0c111236ac2d/gpu/octree/include/pcl/gpu/octree/octree.hpp#L142-L156

</details>

<details>
<summary>detailed logs</summary>

```text
pcl/test/gpu/octree/perfomance.cpp: In member function 'virtual void PCL_OctreeGPU_performance_Test::TestBody()':
pcl/test/gpu/octree/perfomance.cpp:227:72: warning: 'void pcl::gpu::Octree::approxNearestSearch(const Queries&, pcl::gpu::NeighborIndices&) const' is deprecated: use approxNearestSearch() which returns square distances instead (It will be removed in PCL 1.14) [-Wdeprecated-declarations]
         octree_device.approxNearestSearch(queries_device, result_device);
                                                                        ^
In file included from pcl/test/gpu/octree/perfomance.cpp:58:0:
pcl/gpu/octree/include/pcl/gpu/octree/octree.hpp:148:18: note: declared here
             void approxNearestSearch(const Queries& queries, NeighborIndices& result) const;
                  ^~~~~~~~~~~~~~~~~~~
```

</details>

### Type 5: Thrust library issues (tentative)
I'm not very sure about the issue, and didn't find some helpful information. Perhaps add `-Wno-error=maybe-uninitialized` to skip the error?

<details>
<summary>detailed logs</summary>

```text
In file included from tmpxft_0001846c_00000000-5_convex_hull.compute_75.cudafe1.stub.c:1:0:
/tmp/tmpxft_0001846c_00000000-5_convex_hull.compute_75.cudafe1.stub.c:35:1583: error: 'allocations[0]' may be used uninitialized in this function [-Werror=maybe-uninitialized]
 typedef thrust::device_ptr<int>  _ZN6thrust10device_ptrIiEE;
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               ^                                                                                                                                                                                                                                                                                                                                                
/usr/local/cuda/include/thrust/system/cuda/detail/cub/device/dispatch/dispatch_reduce.cuh:511:7: note: 'allocations[0]' was declared here
             void* allocations[1];
       ^     ~~~~~
```

</details>

I've tested all those solutions, and the compile is done successfully in Ubuntu 18.04 GCC. Could you take a look again please? :laughing: (The files changed so far are somehow scattered, maybe it's not very easy for code review. I'd commit the best solutions together after review, in order not to change & push too many times)

<details>

Other things:
<summary>Some false-positive NVCC warnings(won't cause compile to crash with -Werror)</summary>

### Host device functions call host functions (NVCC warnings)

NVCC allows host device functions to call host functions but with a warning being produced. clang provides `-fcuda-allow-host-calls-from-host-device` to disable warning, but gcc doesn't, refered to https://reviews.llvm.org/D7841.

```text
/usr/local/include/eigen3/Eigen/src/Core/arch/CUDA/Half.h(302): warning: calling a __host__ function from a __host__ __device__ function is not allowed
/usr/local/include/eigen3/Eigen/src/Core/arch/CUDA/Half.h(305): warning: calling a __host__ function from a __host__ __device__ function is not allowed
```

### Function declared but never referenced

```text
pcl/gpu/kinfu_large_scale/src/cuda/pointer_shift.cu(4): warning: function "shift_tsdf_pointer" was declared but never referenced
```

</details>

mvieth commented:
Very nice summary! My opinion on the points:
1. I think we should work with the "native" Eigen installation in each Ubuntu release. We will have a CI with Eigen 3.3.9 when we switch 20.10 to 21.04 (although that won't be with GPU). I guess we have to live with those warnings for now
2. I have no opinion on this yet, will try to look into it.
3. Yes, seems like a good idea to make them unsigned
4. Should be fixable just by adding this: `pcl::gpu::Octree::ResultSqrDists sqr_distance;` and changing the call to `octree_device.approxNearestSearch(queries_device, result_device, sqr_distance);`
5. Not sure about this one either. It does not seem to appear on the CI?

Let's not make this pull request much larger - a few more fixes are fine, but if you plan bigger things, maybe do them in a new pull request :smile: There are already many great changes in this PR, let's get it merged soon!

ueqri commented:
Thanks for your kind review and helpful comments! :grinning: 

Should I just skip those tentative warnings, by adding simple rules `-Wno-error=<type>` and some comments as reminders?

> Let's not make this pull request much larger - a few more fixes are fine, but if you plan bigger things, maybe do them in a new pull request.

I plan to dive into the codebase and fix the warnings(which may have risks for future developments). In that case, could I commit changes for the latest summary before this PR being merged? Thank you!

kunaltyagi commented:
> I wonder if there's a plan to move Eigen to the latest stable release 3.3.9, and I'd like to try the migration

As mvieth said, we don't choose the Eigen version, it's the oldest version shipped by the LTS OS we support (See [repology](https://repology.org/project/eigen/versions) for more details)

mvieth commented:
> I plan to dive into the codebase and fix the warnings(which may have risks for future developments). In that case, could I commit changes for the latest summary before this PR being merged? Thank you!

Sure, fixing a few more warning is fine!

ueqri commented:
Thank you both, @kunaltyagi and @mvieth, for your kind support! I'll finally check the changes and push them. :grinning:

kunaltyagi commented on cuda/io/src/debayering.cu:
```diff
@@ -114,8 +114,8 @@ namespace pcl
     	// get position
-    	int xIdx = index % width;
-    	int yIdx = index / width;
+    	unsigned xIdx = index % width;
```
unsigned int?

ueqri replied:
Yes, exactly `unsigned int`. Should I change this to `unsigned int` to make the declaration more explicit?

ueqri replied:
I apologize for being used to declare by a shorthand, which may be not nice for our project.

mvieth commented:
That other noexcept-type warning/error can be solved by using `[](const double& x){ return std::sqrt(x); }` in [test_common.cpp](https://github.com/PointCloudLibrary/pcl/blob/master/test/common/test_common.cpp#L506)

ueqri commented:
Thank you! I'll fix it at once. :smiley:

mvieth reviewed (approved):
Thank you!

kunaltyagi commented on common/include/pcl/common/impl/io.hpp:
```diff
@@ -118,6 +118,25 @@ getFieldsList (const pcl::PointCloud<PointT> &)
 }
 
 
+template <typename PointInT, typename PointOutT> static void
+copyPointCloudMemcpy (const pcl::PointCloud<PointInT> &cloud_in,
+                      pcl::PointCloud<PointOutT> &cloud_out)
+{
+  // Iterate over each point, if the point types of two clouds are different
+  for (std::size_t i = 0; i < cloud_in.size (); ++i)
+    copyPoint (cloud_in[i], cloud_out[i]);
+}
+
+
+template <typename PointT> static void
+copyPointCloudMemcpy (const pcl::PointCloud<PointT> &cloud_in,
+                      pcl::PointCloud<PointT> &cloud_out)
+{
+  // Use std::copy directly, if the point types of two clouds are same
+  std::copy (&cloud_in[0], (&cloud_in[0]) + cloud_in.size (), &cloud_out[0]);
+}
```
Should we move them into a detail namespace? (Technically it's in hpp file, so it should be ok to leave it as it is). Thoughts @mvieth ?

mvieth replied:
Good idea :+1:

ueqri replied:
Thanks for suggesting to move the functions into detail namespace, which really makes the code nicer! :grinning: I've done that.

kunaltyagi reviewed (approved):
LGTM

ueqri commented:
I'm really grateful for all your kind suggestions and helps in this PR! I learned so much through the process of tackling every problems I met to fix the warnings and fine-tune the codes.

I'm quite willing to make more contributions to fixing underlying warnings and code modernization using `clang-tidy` in the next steps. This is indeed a pleasure for me! :grinning:
